base/system/winlogon/security.c

```diff
@@ -0,0 +1,1490 @@
+    {
+        RtlFreeHeap(RtlGetProcessHeap(), 0, Dacl);
+    }
+
+    if (RelativeSd != NULL)
+    {
+        RtlFreeHeap(RtlGetProcessHeap(), 0, RelativeSd);
+    }
+
+    return Success;
+}
+
+/**
+ * @brief
+ * Assigns both window station and desktop access
+ * to the specific session currently active on the
+ * system.
+ *
+ * @param[in] Session
+ * A pointer to an active session.
+ *
+ * @return
+ * Returns TRUE if the function has successfully
+ * assigned access to the current session, FALSE otherwise.
+ */
+BOOL
+AllowAccessOnSession(
+    _In_ PWLSESSION Session)
+{
+    BOOL Success = FALSE;
+    DWORD Index, SidLength, GroupsLength = 0;
+    PTOKEN_GROUPS TokenGroup = NULL;
+    PSID LogonSid;
+
+    /* Get required buffer size and allocate the TOKEN_GROUPS buffer */
+    if (!GetTokenInformation(Session->UserToken,
+                             TokenGroups,
+                             TokenGroup,
+                             0,
+                             &GroupsLength))
+    {
+        if (GetLastError() != ERROR_INSUFFICIENT_BUFFER)
+        {
+            ERR("AllowAccessOnSession(): Unexpected error code returned, must be ERROR_INSUFFICIENT_BUFFER (error code %lu)\n", GetLastError());
+            return FALSE;
+        }
+
+        TokenGroup = RtlAllocateHeap(RtlGetProcessHeap(), HEAP_ZERO_MEMORY, GroupsLength);
+        if (TokenGroup == NULL)
+        {
+            ERR("AllowAccessOnSession(): Failed to allocate memory buffer for token group!\n");
+            return FALSE;
+        }
+    }
+
+    /* Get the token group information from the access token */
+    if (!GetTokenInformation(Session->UserToken,
+                             TokenGroups,
+                             TokenGroup,
+                             GroupsLength,
+                             &GroupsLength))
+    {
+        ERR("AllowAccessOnSession(): Failed to retrieve the token group information (error code %lu)\n", GetLastError());
+        goto Quit;
+    }
+
+    /* Loop through the groups to find the logon SID */
+    for (Index = 0; Index < TokenGroup->GroupCount; Index++)
+    {
+        if ((TokenGroup->Groups[Index].Attributes & SE_GROUP_LOGON_ID)
+            == SE_GROUP_LOGON_ID)
+        {
+            LogonSid = TokenGroup->Groups[Index].Sid;
+            break;
+        }
+    }
+
+    /* Allow window station access to this user within this session */
+    if (!AllowWinstaAccessToUser(Session->InteractiveWindowStation, LogonSid))
+    {
+        ERR("AllowAccessOnSession(): Failed to allow winsta access to the logon user!\n");
+        goto Quit;
+    }
+
+/*
+ * FIXME: Desktop security management is broken. The application desktop gets created
+ * with CreateDesktopW() API call on wlx.file with an initial and defined security
+ * descriptor for it yet when we are giving access to the logged in user, SetUserObjectSecurity()
+ * API call fails to set new security because the desktop in question has no prior security
```

Comment:
Do you know where the bug may reside?

Reply:
No, I am not sure nor I am aware where's the root problem of this bug yet, I will look into this in the future when I can, for now this code path has to stay disabled (otherwise you ROS won't be able to log in anymore).